harness/determined/layers/_workload_manager.py

```diff
@@ -274,59 +308,130 @@ def _respond(in_response: workload.Response) -> None:
             respond(out_response)
 
+
+        start_time = _current_timestamp()
+        if not self.env.test_mode:
+            validation_metrics = {
+                "trial_id": self.env.get_trial_id(),
+                "total_batches": wkld.total_batch(),
+                "start_time": start_time.strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
+                "state": "STATE_ACTIVE",
+            }
+            req.post(
+                self.env.get_unary_host(),
+                "/api/v1/validation_metrics",
+                body={"validation_metrics": validation_metrics},
+            )
+
         for callback in self.callbacks:
             callback.on_validation_step_begin(wkld.step_id, wkld.total_batches_processed)
 
+        # Yield the workload to the framework layer.
         yield wkld, [], _respond
 
     def yield_checkpoint_model(
         self, wkld: workload.Workload, respond: workload.ResponseFunc
     ) -> workload.Stream:
-        start_time = _current_timestamp()
-
-        # Only the chief container should checkpoint.
-        if self.rendezvous_info.get_rank() != 0:
-            respond(workload.Skipped())
-            return
 
         # Save the workload completed message for after checkpoint upload completes.
         message = None  # type: Optional[workload.Response]
 
         def _respond(checkpoint_info: workload.Response) -> None:
             checkpoint_info = cast(Dict[str, Any], checkpoint_info)
+            resources = storage.StorageManager._list_directory(path)
+            framework = checkpoint_info.get("framework", "")
+            format = checkpoint_info.get("format", "")
+            end_time = _current_timestamp()
+
             metadata = storage.StorageMetadata(
-                storage_id,
-                storage.StorageManager._list_directory(path),
-                checkpoint_info.get("framework", ""),
-                checkpoint_info.get("format", ""),
+                uuid,
+                resources,
+                framework,
+                format,
             )
 
-            logging.info("Saved trial to checkpoint {}".format(metadata.storage_id))
+            if not self.env.test_mode:
+                checkpoint = {
+                    "trial_id": self.env.get_trial_id(),
+                    "batch_number": wkld.total_batch(),
+                    "start_time": start_time.strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
+                    "end_time": end_time.strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
+                    "state": "STATE_COMPLETED",
+                    "uuid": uuid,
+                    "resources": resources,
+                    "metadata": metadata.__dict__,
+                    "format": format,
+                    "framework": framework,
+                    "determined_version": det.__version__,
+                }
+                req.put(
+                    self.env.get_unary_host(), "/api/v1/checkpoints", body={"checkpoint": checkpoint}
+                )
```

Comment:
This is a bug: the checkpoint is being reported before it is being actually uploaded.  This `put` must occur after the `with self.storage_mgr.store_path() as (uuid, path):` context manager exits.

(I only recognize this bug because I made the exact same mistake before : )

Reply:
: ) thanks for pointing it out! This makes me hate the response function more.